test/integration/metrics_test.go

```diff
@@ -0,0 +1,259 @@
+	clock := fakeclock.NewFakeClock(time.Now())
+	opts, cl, stop := testdriver.Run(t, testdriver.Options{
+		Store:   store,
+		Clock:   clock,
+		Metrics: metricsHandler,
+		Log:     &testLog,
+		GeneratePrivateKey: func(meta metadata.Metadata) (crypto.PrivateKey, error) {
+			return nil, nil
+		},
+		GenerateRequest: func(meta metadata.Metadata) (*manager.CertificateRequestBundle, error) {
+			return &manager.CertificateRequestBundle{
+				Namespace: testNamespace,
+				IssuerRef: cmmeta.ObjectReference{
+					Name:  "test-issuer",
+					Kind:  "test-issuer-kind",
+					Group: "test-issuer-group",
+				},
+			}, nil
+		},
+		SignRequest: func(meta metadata.Metadata, key crypto.PrivateKey, request *x509.CertificateRequest) (csr []byte, err error) {
+			return []byte{}, nil
+		},
+		WriteKeypair: func(meta metadata.Metadata, key crypto.PrivateKey, chain []byte, ca []byte) error {
+			store.WriteFiles(meta, map[string][]byte{
+				"ca":   ca,
+				"cert": chain,
+			})
+			nextIssuanceTime := clock.Now().Add(time.Hour)
+			meta.NextIssuanceTime = &nextIssuanceTime
+			return store.WriteMetadata(meta.VolumeID, meta)
+		},
+	})
+	defer stop()
+
+	// Should expose no additional metrics
+	metricsEndpoint := fmt.Sprintf("http://%s/metrics", metricsServer.Addr)
+	waitForMetrics(t, ctx, metricsEndpoint, "")
+
+	// Create a self-signed Certificate and wait for it to be issued
+	privKey := testcrypto.MustCreatePEMPrivateKey(t)
+	certTemplate := &cmapi.Certificate{
+		ObjectMeta: metav1.ObjectMeta{Namespace: testNamespace, Name: "test"},
+		Spec: cmapi.CertificateSpec{
+			CommonName: "test.example.com",
+		},
+	}
+	notBefore, notAfter := time.Unix(0, 0), time.Unix(300, 0) // renewal time will be 200
+	selfSignedCertBytesWithValidity := testcrypto.MustCreateCertWithNotBeforeAfter(t, privKey, certTemplate, notBefore, notAfter)
+	go testutil.IssueOneRequest(ctx, t, opts.Client, testNamespace, selfSignedCertBytesWithValidity, []byte("ca bytes"))
+
+	// Spin up a test pod
+	tmpDir, err := os.MkdirTemp("", "*")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(tmpDir)
+	_, err = cl.NodePublishVolume(ctx, &csi.NodePublishVolumeRequest{
+		VolumeId: "test-vol",
+		VolumeContext: map[string]string{
+			"csi.storage.k8s.io/ephemeral":     "true",
+			"csi.storage.k8s.io/pod.name":      "the-pod-name",
+			"csi.storage.k8s.io/pod.namespace": testNamespace,
+		},
+		TargetPath: tmpDir,
+		Readonly:   true,
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Get the CSR name
+	req, err := testutil.WaitAndGetOneCertificateRequestInNamespace(ctx, opts.Client, testNamespace)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Should expose that CertificateRequest as ready with expiry and renewal time
+	// node="f56fd9f8b" is the hash value of "test-node" defined in driver_testing.go
+	expectedOutputTemplate := `# HELP certmanager_csi_certificate_request_expiration_timestamp_seconds The date after which the certificate request expires. Expressed as a Unix Epoch Time.
```

Comment:
Will `certmanager_csi_...` be the prefix for the metrics when exposed in https://github.com/cert-manager/csi-driver?
Can we make sure they are all using similar names? Does this happen automatically?
Could you provide an example of before and after for https://github.com/cert-manager/csi-driver with these changes applied?

Reply:
Thank you @inteon for your review.

Yes, `certmanager_csi_...` will be the prefix for the metrics when in the csi-driver. It is part of the definition from:
https://github.com/7ing/csi-lib/blob/b9186bad5b6f9af9bc93dbd28571d2d9219700e6/metrics/metrics.go#L27-L31

We choose this name based on cert-manager.io definition: https://github.com/cert-manager/cert-manager/blob/5e09ef6c0552df0bde64746c735cb1ff324b6261/pkg/metrics/metrics.go#L44
All cert-manager controllers have `certmanager_..` prefix.

Our https://github.com/cert-manager/csi-driver does not have any certmanager related metrics. Currently it only serves k8s components metrics, like cpu / mem etc. That's why this PR exist.